client/asset/interface.go

```diff
@@ -1272,3 +1271,79 @@ type MultiOrder struct {
 	// RedeemAssetID is the asset ID of the "to" asset.
 	RedeemAssetID uint32
 }
+
+// WalletNotification can be any asynchronous information the wallet needs
+// to convey.
+type WalletNotification interface{}
+
+// TipChangeNote is the only required wallet notification. All wallets should
+// emit a TipChangeNote when a state change occurs that might necessitate swap
+// progression or new balance checks.
+type TipChangeNote struct {
+	AssetID uint32      `json:"assetID"`
+	Tip     uint64      `json:"tip"`
+	Data    interface{} `json:"data"`
+}
+
+// AsyncWalletErrorNote is issued when the wallet detects it is in an error
+// state.
+type AsyncWalletErrorNote struct {
+	AssetID uint32 `json:"assetID"`
+	Err     error  `json:"err"`
+}
+
+// CustomWalletNote is any other information the wallet wishes to convey to
+// the user.
+type CustomWalletNote struct {
+	AssetID uint32      `json:"assetID"`
+	Payload interface{} `json:"payload"`
+}
+
+// WalletEmitter handles a channel for wallet notifications and provides methods
+// that generates notifications.
+type WalletEmitter struct {
+	c       chan<- WalletNotification
+	assetID uint32
+	log     dex.Logger
+}
+
+// NewWalletEmitter constructs a WalletEmitter for an asset.
+func NewWalletEmitter(c chan<- WalletNotification, assetID uint32, log dex.Logger) *WalletEmitter {
+	return &WalletEmitter{
+		c:       c,
+		assetID: assetID,
+		log:     log,
+	}
+}
+
+func (e *WalletEmitter) emit(note WalletNotification) {
+	select {
+	case e.c <- note:
+	default:
+		e.log.Warn("blocking WalletNotification channel")
+	}
+}
+
+// Data sends a CustomWalletNote with the specified data payload.
+func (e *WalletEmitter) Data(payload interface{}) {
+	e.emit(&CustomWalletNote{AssetID: e.assetID, Payload: payload})
+}
+
+// Errorf formats and sends an AsyncWalletErrorNote.
+func (e *WalletEmitter) Errorf(s string, a ...interface{}) {
```

Comment:
Instead of only `Errorf`, a general `Logf` note with an error level argument would be useful to let `core` know the severity of the notification to send to the front end.

Reply:
Did you have a particular place you were thinking it was needed, because I was thinking of removing `Errorf` and `Error` altogether. The only thing `Core` does is log it, and the wallet already has a logger that logs to the same place but with more context. We can add `Logf` or anything else as needed, but I don't have a use for it right now.

Reply:
Yeah it's definitely not needed yet.